File: src/app/stats.py
from registro_iphone import registro_iphone
from registro import registro
import datetime

class stats:

    #Dentro de la variable numMensajes se guardara la informacion y el tipo de cada mensaje
    """
    Esta lista recogera cuantos mensajes de cada tipo se han enviado, 
    de esta forma los tipos de mensajes y su indice en la lista son los siguientes

        Mensajes de texto--------------- [0]
        Mensajes de Audio--------------- [1]
        Fotos--------------------------- [2]
        Videos-------------------------- [3]
        Stickers------------------------ [4]
        gifs---------------------------- [5]
        Multimedia omitido-------------- [6]

    Al exportar un chat en dispositivos android no sera posible distinguir entre 
    audios,fotos,videos etc... Todo ira englobado en Multimedia omitido
    """ 
    
    """
    
    Para Gestionar los mensajes por meses se ha creado un segundo diccionario que tendra como claves
    los años en los que ha habido conversacion.

    Enero------------------------------ [0]
    Febrero---------------------------- [1]
    Marzo------------------------------ [2]
    Abril------------------------------ [3]
    Mayo------------------------------- [4]
    Junio------------------------------ [5]
    Julio------------------------------ [6]
    Agosto----------------------------- [7]
    Septiembre------------------------- [8]
    octubre---------------------------- [9]
    Noviembre-------------------------- [10]
    Diciembre-------------------------- [11]

    """



    def __init__(self,archivo,opcion):
       
        #Esta lista recogera todas las personas que participan en la conversacion

        #Nombre de la persona----------- [0]
        #numMensajes[] ----------------- [1]

       self.datos = {}

       if opcion == 1:
        self.recorrer_fichero_android(archivo)
       else:
        self.recorrer_fichero_iphone(archivo)

    def recorrer_fichero_android(self,filename):
        with open(filename, 'r', encoding="utf-8") as f:
            next(f)
            for linea in f:   
               try:
                    r = registro(linea)
                    nombre=r.getPersona()

                    if not (nombre in self.datos):
                        self.añadir_persona(nombre)

                    self.clasificar_mensaje(nombre,r.getMensaje())

                    #self.clasificar_tiempo(nombre,r.getFechaHora)
               except:
                   continue 

            print(self.datos) 

    def recorrer_fichero_iphone(self,filename):
        with open(filename, 'r', encoding="utf-8") as f:
            next(f)
            for linea in f:   
               try:
                    linea = linea.encode("ascii","ignore")
                    linea = linea.decode()
                    linea = linea.strip()
                    
                    r = registro_iphone(linea)
                    nombre=r.get_persona()
                    mensaje = r.get_mensaje()
                    fechaHora = r.get_fecha_hora()

                    if not (nombre in self.datos):
                        self.añadir_persona(nombre)

                    self.clasificar_mensahe_iphone(nombre,mensaje)
                    self.clasificar_tiempo(nombre,fechaHora)
                    #print(self.datos)

               except:
                   continue     
            print(self.datos)      
            

    def clasificar_tiempo(self,nombre,date):

        if not (date.year in self.datos[nombre][1]):
            self.datos[nombre][1][date.year] =[0,0,0,0,0,0,0,0,0,0,0,0]
        

        match date.month:

            case 1:
                self.datos[nombre][1][date.year][0] += 1
            case 2:
                self.datos[nombre][1][date.year][1] += 1
            case 3:
                self.datos[nombre][1][date.year][2] += 1
            case 4:
                self.datos[nombre][1][date.year][3] += 1
            case 5:
                self.datos[nombre][1][date.year][4] += 1
            case 6:
                self.datos[nombre][1][date.year][5] += 1
            case 7:
                self.datos[nombre][1][date.year][6] += 1
            case 8:
                self.datos[nombre][1][date.year][7] += 1
            case 9:
                self.datos[nombre][1][date.year][8] += 1
            case 10:
                self.datos[nombre][1][date.year][9] += 1
            case 11:
                self.datos[nombre][1][date.year][10] += 1
            case 12:
                self.datos[nombre][1][date.year][11] += 1

            
    def clasificar_mensaje(self,nombre,mensaje):

        if mensaje.strip() == "<Multimedia omitido>":
            self.datos[nombre][0][6] += 1
        elif mensaje.strip() == "sticker omitido":
            self.datos[nombre][0][4] += 1
        elif mensaje.strip() == "imagen omitida":
            self.datos[nombre][0][2] += 1
        elif mensaje.strip() == "Video omitido":
            self.datos[nombre][0][3] += 1
        else:
            self.datos[nombre][0][0] += 1

    def clasificar_mensahe_iphone(self,nombre,mensaje):
        
        if mensaje == "sticker omitido":
            self.datos[nombre][0][4] += 1
        elif mensaje == "imagen omitida":
            self.datos[nombre][0][2] += 1
        elif mensaje == "Video omitido":
            self.datos[nombre][0][3] += 1
        elif mensaje == "audio omitido":
            self.datos[nombre][0][1] += 1
        else:
            self.datos[nombre][0][0] += 1

        
    
    def añadir_persona(self,nombre):

        numMensajes = [0,0,0,0,0,0,0]
        meses={}
        
        info=[numMensajes,meses]

        self.datos[nombre] = info

        
            



        

File: src/fichero/Stats.ts
/* eslint-disable @typescript-eslint/no-unused-vars */

import RegistroIphone from "./registroIphone";
import {
  FechaHora,
  MensajesType,
  mesesMap,
  Registro,
  Meses,
} from "./types";

class Stats {
  private fichero: File | null;
  private dispositivo: string;
  private stats: Registro[];
  private ficheroLeido;

  constructor(fichero: File | null, dispositivo: string) {
    this.fichero = fichero;
    this.dispositivo = dispositivo;
    this.stats = [];
    this.ficheroLeido = false;
  }

  async analizar() {
    //Se lee el fichero
    await this.leerFichero();
  }

  private leerFichero(): Promise<void> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();

      reader.onloadend = () => {
        this.ficheroLeido = true;
        resolve();
      };

      reader.onload = (e) => {
        let text = e.target?.result as string;

        // Se eliminan caracteres no deseados como caracteres invisibles
        text = this.eliminarCaracteresNoDeseados(text);

        // Se separa el texto en líneas
        const lines = text
          .split(/\r?\n/)
          .map((linea) => linea.replace(/\r/g, ""));

        let errors = "";

        if (this.dispositivo == "iphone") {
          lines.forEach((line) => {
            try {
              const registroIphone = new RegistroIphone(line);
              const personaRegistro = registroIphone.getPersona();
              const mensajeRegistro = registroIphone.getMensaje();
              const fechaHoraRegistro = registroIphone.getFechaHora();
              const año = fechaHoraRegistro.fecha.split("/")[2];

              const añadirPersona = this.stats.some(
                (persona) =>
                  persona.nombre === personaRegistro && persona.año === año,
              );

              if (!añadirPersona) {
                this.añadirPersona(personaRegistro, año);
              }

              this.clasificarMensaje(
                personaRegistro,
                fechaHoraRegistro,
                mensajeRegistro,
              );
            } catch (err) {
              errors += line + "\n";
            }
          });
        } else {
          lines.forEach((line) => {
            // Lógica para Android aquí...
          });
        }
      };

      reader.onerror = (e) => {
        reject(e);
      };

      if (this.fichero) {
        reader.readAsText(this.fichero, "utf-8");
      } else {
        console.error("No se ha definido ningún archivo para leer.");
        reject(new Error("No se ha definido ningún archivo para leer."));
      }
    });
  }
  //Se eliminan caracteres no Deseados como caracteres invisibles
  private eliminarCaracteresNoDeseados(texto: string): string {
    return texto.replace(/[\r\u200E]/g, "");
  }

  //Se añade una persona al Array de Stats
  private añadirPersona(personaNombre: string, año: string) {
    //Se crea un objeto persona con su nombre
    const personaObjeto: Registro = {
      nombre: personaNombre,
      año: año,
      mensajes: [this.añadirMeses()],
    };
    //Se añade al array
    this.stats.push(personaObjeto);
  }

  //Se clasifica el mensaje por tiempo y tipo de mensaje
  private clasificarMensaje(
    persona: string,
    fechaHora: FechaHora,
    mensajeRegistro: string,
  ) {
    //Separamos las variables que nos interesan de fechaHora
    const mesAux = fechaHora.fecha.split("/")[1];

    const añoRegistro = fechaHora.fecha.split("/")[2];
    const mesRegistro = this.numeroAMes(Number(mesAux));
    const diaRegistro = fechaHora.fecha.split("/")[0];

    const aux = this.stats.find(
      (registro) => registro.año === añoRegistro && registro.nombre === persona,
    );

    aux?.mensajes.map((año) => {
      let mensajesData;

      if (mesRegistro) {
        mensajesData = año[mesRegistro as keyof Meses];

        //Normalizamos el mensaje para evitar espacios en blanco o mayusculas no deseadas
        const mensajeNormalizado = mensajeRegistro
          .trim()
          .toLowerCase()
          .valueOf();

        //Se identifica el tipo de mensaje y se suma 1 al array
        switch (mensajeNormalizado) {
          case "sticker omitido":
            mensajesData.stickers += 1;
            break;
          case "audio omitido":
            mensajesData.mensajesAudio += 1;
            break;
          case "Video omitido":
            mensajesData.videos += 1;
            break;
          case "imagen omitida":
            mensajesData.fotos += 1;
            break;
          case "gif omitido":
            mensajesData.gifs += 1;
            break;
          default:
            mensajesData.mensajesTexto += 1;
            break;
        }
      }
    });
  }

  //Crea una instancia de MensajeType en blanco
  private crearMensajesType = (): MensajesType => ({
    mensajesTexto: 0,
    mensajesAudio: 0,
    fotos: 0,
    videos: 0,
    stickers: 0,
    gifs: 0,
    multimediaOmitido: 0,
  });

  //Crea un objeto Año con los meses y dentro de ellos la estructura de MensajeType
  private añadirMeses() {
    const añoData: Meses = {
      enero: this.crearMensajesType(),
      febrero: this.crearMensajesType(),
      marzo: this.crearMensajesType(),
      abril: this.crearMensajesType(),
      mayo: this.crearMensajesType(),
      junio: this.crearMensajesType(),
      julio: this.crearMensajesType(),
      agosto: this.crearMensajesType(),
      septiembre: this.crearMensajesType(),
      octubre: this.crearMensajesType(),
      noviembre: this.crearMensajesType(),
      diciembre: this.crearMensajesType(),
    };
    return añoData;
  }


  //Devuelve el nombre de un mes a partir de su numero de mes
  private numeroAMes(numero: number): string | undefined {
    return mesesMap[numero];
  }

  public obtenerData() {
    return JSON.stringify(this.stats);
  }
}

export default Stats;
